test/integration/dualstack/dualstack_test.go

```diff
@@ -0,0 +1,1386 @@
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+
+			svc := &v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: fmt.Sprintf("svc-test-%d", i), // use different services for each test
+				},
+				Spec: v1.ServiceSpec{
+					Type:           tc.serviceType,
+					ClusterIPs:     tc.clusterIPs,
+					IPFamilies:     tc.ipFamilies,
+					IPFamilyPolicy: &tc.ipFamilyPolicy,
+					Ports: []v1.ServicePort{
+						{
+							Name:       fmt.Sprintf("port-test-%d", i),
+							Port:       443,
+							TargetPort: intstr.IntOrString{IntVal: 443},
+							Protocol:   "TCP",
+						},
+					},
+				},
+			}
+
+			// create a service
+			_, err = client.CoreV1().Services(metav1.NamespaceDefault).Create(context.TODO(), svc, metav1.CreateOptions{})
+			if (err != nil) != tc.expectError {
+				t.Errorf("Test failed expected result: %v received %v ", tc.expectError, err)
+			}
+			// if no error was expected validate the service otherwise return
+			if err != nil {
+				return
+			}
+			// validate the service was created correctly if it was not expected to fail
+			svc, err = client.CoreV1().Services(metav1.NamespaceDefault).Get(context.TODO(), svc.Name, metav1.GetOptions{})
+			if err != nil {
+				t.Errorf("Unexpected error to get the service %s %v", svc.Name, err)
+			}
+
+			ipFamilies := tc.ipFamilies
+			// if no IPFamily specified it must use the defaults (first cidrs in service subnet)
+			if (tc.ipFamilyPolicy == v1.SingleStack) && (len(ipFamilies) == 0) {
+				ipFamilies = defaultIPFamily[:1]
+				// if no IPFamily specified it must use the defaults
+			} else if (tc.ipFamilyPolicy != v1.SingleStack) && (len(ipFamilies) == 0) {
+				ipFamilies = defaultIPFamily
+				// if only one IPFamily specified but dual stack add the missing one
+			} else if (tc.ipFamilyPolicy != v1.SingleStack) && (len(ipFamilies) == 1) {
+				if ipFamilies[0] == v1.IPv4Protocol {
+					ipFamilies = append(ipFamilies, v1.IPv6Protocol)
+				} else {
+					ipFamilies = append(ipFamilies, v1.IPv4Protocol)
+				}
+			}
+			if err := validateServiceAndClusterIPFamily(svc, ipFamilies); err != nil {
+				t.Errorf("Unexpected error validating the service %s %v", svc.Name, err)
+			}
+		})
+	}
+}
+
+// helper functions obtained from test/e2e/network/dual_stack.go
+
+// validateServiceAndClusterIPFamily checks that the service has the expected IPFamilies
+func validateServiceAndClusterIPFamily(svc *v1.Service, expectedIPFamilies []v1.IPFamily) error {
+	// create a slice for the errors
+	var errstrings []string
+
+	if svc.Spec.IPFamilies == nil {
+		return fmt.Errorf("service ip family nil for service %s/%s", svc.Namespace, svc.Name)
+	}
+	if !reflect.DeepEqual(svc.Spec.IPFamilies, expectedIPFamilies) {
+		return fmt.Errorf("ip families mismatch for service: %s/%s, expected: %s, actual: %s", svc.Namespace, svc.Name, expectedIPFamilies, svc.Spec.IPFamilies)
+	}
+
+	for j, ip := range svc.Spec.ClusterIPs {
+		if ip == v1.ClusterIPNone {
+			// ???
```

Comment:
I forget if we decided this is allowed or not, but if it is, it would be an error in these tests